### PR TITLE
Windows, docs: tips and tricks to write rules

### PR DESCRIPTION
**Pull request by laszlocsomor**
See https://github.com/bazelbuild/bazel/issues/3889

Change-Id: Id768d8bd0cb996af566cdda518772ea329e5c6b2

### CODE REVIEW

File: site/docs/skylark/windows_tips.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  Processes cannot `cd` into a directory longer than 259 characters.
+
+- **Case-sensitivity**: Windows paths are case-insensitive, Unix paths are case-sensitive.
+
+  Be aware of this when creating command lines for actions.
+
+- **Path separators**: are backslash (`\`), not forward slash (`/`).
+
+  Bazel uses paths with `/` separators, so on Windows you have to replace those with `\` when
+  creating comamnd lines and environment variables for actions.
```
Comment (laurentlb) [inline site/docs/skylark/windows_tips.md]: In many cases `/` works on Windows, from what I remember. Do you know more about it?

File: site/docs/skylark/windows_tips.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+- **Absolute paths**: don't start with slash (`/`).
+
+  Absolute paths on Windows start with a drive letter, e.g. `C:\foo\bar.txt`. There's no single
+  filesystem root.
+
+  Be aware of this if your rule checks if a path is absolute.
```
Comment (laurentlb) [inline site/docs/skylark/windows_tips.md]: In general, absolute paths should be avoided though. Absolute paths are often non-portable.

File: site/docs/skylark/windows_tips.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  ```
+  mklink /J c:\path\to\junction c:\path\to\very\long\target\path
+  ```
+
+- **Replace `/` with `\` in paths in actions / envvars.**
+
+  When you create the command line or environment variables for an action, make the paths
+  Windows-style. Example:
+
+  ```
```
Comment (laurentlb) [inline site/docs/skylark/windows_tips.md]: For syntax highlighting, use:
```
```python
```

File: site/docs/skylark/windows_tips.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+  error code, a simple `.bat` file will often suffice. And if your rule returns a `DefaultInfo()`
+  provider, the `executable` field may refer to that `.bat` file on Windows.
+
+  And since file extensions don't matter on macOS / Linux, you can always use `.bat` as the
+  extension.
+
+  Be aware that **empty `.bat` files cannot be executed**. If you need an empty script, write one
+  space in it.
+
+- **Use a Bash-less purpose-made rule.**
```
Comment (laurentlb) [inline site/docs/skylark/windows_tips.md]: Move this section before the advice "use bat scripts".

File: site/docs/skylark/windows_tips.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+- **Use Bash in a principled way.**
+
+  You can still run Bash commands if you need to, just let Bazel know that it's a Bash command.
+
+  In Starlark rules:
+
+  - Use `ctx.actions.run_shell` if you need to run Bash commands.
+
+  - **Do not** use `ctx.actions.run` with Bash commands -- this won't work on Windows because Bazel
+    doesn't know that it needs to run the command through Bash.
```
Comment (laurentlb) [inline site/docs/skylark/windows_tips.md]: What do you mean? `ctx.actions.run` doesn't use Bash.

File: site/docs/skylark/windows_tips.md (hunk 1 above)
Comment (laszlocsomor) [reply site/docs/skylark/windows_tips.md]: Indeed. `/` support is inconsistent.

It works when the binary you run supports it. (Typically those binaries use the ANSI subset of the Windows API. They are also the ones that don't support long paths.)

Built-in cmd.exe commands (like `copy`, `del`) tend not to support it:
```
C:\foo>copy bar/baz/qux bar/baz/qox
The syntax of the command is incorrect.

C:\foo>copy bar\baz\qux bar/baz/qox
The syntax of the command is incorrect.

C:\foo>copy bar\baz\qux bar\baz\qox
        1 file(s) copied.
```

but sometimes they do (e.g. when piping):

```
C:\foo>copy bar\baz\qux bar\baz\qox
        1 file(s) copied.

C:\foo>echo x > bar/baz/qaax

C:\foo>type bar/baz/qaax
The syntax of the command is incorrect.

C:\foo>type bar\baz\qaax
x
```

Comment (laurentlb) [reply site/docs/skylark/windows_tips.md]: Maybe put one line about it.

**Comment by laszlocsomor** [comment]
PTAL

File: site/docs/skylark/windows_tips.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  the files in the junction'ed directory.
+
+  In `.bat` files or in cmd.exe you can create junctions like so:
+
+  ```
+  mklink /J c:\path\to\junction c:\path\to\very\long\target\path
+  ```
+
+  <sup>[1]</sup>: Strictly speaking
+  [Junctions are not Symbolic Links](https://superuser.com/a/343079), but for sake of build actions
```
Comment (meteorcloudy) [inline site/docs/skylark/windows_tips.md]: Maybe we should have a individual section about "symlink", we can explain why runfiles tree doesn't exist by default on Windows, and what are the solutions (runfiles library / --enable_runfiles)

File: site/docs/skylark/windows_tips.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    ([source](https://github.com/bazelbuild/bazel-skylib/blob/master/rules/diff_test.bzl),
+    [documentation](https://github.com/bazelbuild/bazel-skylib/blob/master/docs/diff_test_doc.md)):
+    test that compares contents of two files
+
+  - `native_test()`
+    ([source](https://github.com/bazelbuild/bazel-skylib/blob/master/rules/native_binary.bzl),
+    [documentation](https://github.com/bazelbuild/bazel-skylib/blob/master/docs/native_binary_doc.md#native_test)):
+    wraps a native binary in a `*_test` rule, which you can `bazel test`
+
+- **On Windows, consider using `.bat` scripts for trivial things.**
```
Comment (meteorcloudy) [inline site/docs/skylark/windows_tips.md]: Maybe we could also mention using `powershell` as a solution?

**Comment by meteorcloudy** [comment]
This will be amazingly helpful! Thanks for writing this!

File: site/docs/skylark/windows_tips.md (hunk 6 above)
Comment (laszlocsomor) [reply site/docs/skylark/windows_tips.md]: Absolutely! Runfiles are a top TODO, see comments at the bottom of the file.

File: site/docs/skylark/windows_tips.md (hunk 7 above)
Comment (laszlocsomor) [reply site/docs/skylark/windows_tips.md]: Good idea!

But I'm reluctant to do so. Running `.ps` scripts is harder than `.bat`: you can pass a `.bat` file to `CreateProcessW` directly (therefore also to `ctx.actions.run(executable={})`, but not a `.ps` file.

**Comment by laszlocsomor** [comment]
Thank you! And for the reviews too :)

File: site/docs/skylark/windows_tips.md (hunk 6 above)
Comment (meteorcloudy) [reply site/docs/skylark/windows_tips.md]: Nice! I was reviewing the rendered markdown page, so I missed it.

File: site/docs/skylark/windows_tips.md (hunk 7 above)
Comment (meteorcloudy) [reply site/docs/skylark/windows_tips.md]: I see, if we cannot run .ps directly when starting a process, then I think it's fine to only mention batch! Thanks!

**Pull request closed without merging**